packages/apollo-server-plugin-base/src/index.ts

```diff
@@ -38,27 +44,31 @@ export {
   GraphQLRequestContextWillSendResponse,
 };
 
-export interface ApolloServerPlugin<TContext extends Record<string, any> = Record<string, any>> {
+export interface ApolloServerPlugin<
+  TContext extends BaseContext = DefaultContext
+> {
   serverWillStart?(service: GraphQLServiceContext): ValueOrPromise<void>;
   requestDidStart?(
     requestContext: GraphQLRequestContext<TContext>,
   ): GraphQLRequestListener<TContext> | void;
 }
 
-export type GraphQLRequestListenerParsingDidEnd =
-  ((err?: Error) => void) | void;
+export type GraphQLRequestListenerParsingDidEnd = (err?: Error) => void;
 export type GraphQLRequestListenerValidationDidEnd =
-  ((err?: ReadonlyArray<Error>) => void) | void;
-export type GraphQLRequestListenerExecutionDidEnd =
-  ((err?: Error) => void) | void;
+  ((err?: ReadonlyArray<Error>) => void);
+export type GraphQLRequestListenerExecutionDidEnd = ((err?: Error) => void);
+export type GraphQLRequestListenerDidResolveField =
+  ((error: Error | null, result?: any) => void);
 
-export interface GraphQLRequestListener<TContext = Record<string, any>> {
+export interface GraphQLRequestListener<
+  TContext extends BaseContext = DefaultContext
+> extends AnyFunctionMap {
   parsingDidStart?(
     requestContext: GraphQLRequestContextParsingDidStart<TContext>,
-  ): GraphQLRequestListenerParsingDidEnd;
+  ): GraphQLRequestListenerParsingDidEnd | void;
```

Comment:
I think these should probably be `| undefined` rather than `| void`. (The uses of `=> void` without the `|` in the definitions of the types above seem correct, though.)

See eg [this comparison of `void` with `undefined` from the TS engineering lead](https://stackoverflow.com/a/58885486).  If the idea of `void` is "I won't look at the return value from this callback", then `| void` doesn't make much sense, because you do have to look at the return value in order to decide which side of the `|` is valid.

(I might be misunderstanding — this is my introduction to learning about TS `void`!)

Reply:
I might agree with you, but I think that'll be better resolved in another PR rather than under the umbrella of this one; This is a pattern that existed before this PR and before the recent https://github.com/apollographql/apollo-server/pull/3985 which shuffled their definition location but didn't change their values.

Tangentially though: one interpretation of this type might be, "You can return a function that accepts an optional error and returns a value which we will not care about, **or** you can return a value that we will not care about."

I do believe that we have to assess whether the value is a function in order to determine if we can invoke it, but I'm not certain if that violates the contract of observation.

Anyhow, let's discuss this, but not have it block this?

Reply:
Agreed that it isn't salient to this particular PR, though you may want to file a ticket and link it here or at least add a comment. Fine if you don't want to though; you know this project's standards better than I.

I am curious to see if my comment on the StackOverflow above gets a response.

I am not an expert but your proposed interpretation does not really ring true to me; it really feels like "not caring about the return value of a function" isn't a half-way thing. Or well, I can imagine `f(x: boolean): number | void` meaning "returns a number if `x == true`, ignores return value if `x == false`", but not just "returns a value that you have to check before deciding to ignore it.

I recognize that this isn't a backwards-incompatible change though, right? Like, the old one could return a number and the new one can't? Probably good to be stricter if we feel comfortable with that since that lets us feel better about expanding legal return types later.

Anyway, not this PR.

Reply:
Yeah, it was just a thought.  In some ways, it feels like we're just doing a structural type check at runtime, but I certainly get your perspective and I think that specifying `undefined` here would prevent unintentional expectations (though possibly thwart anyone using shorthand fat-arrow notation that syntactic-sugarly returns the value on its right (e.g. `=> me()`).  I up-⛵'d your comment and hope we get some thought.

Also, logged https://github.com/apollographql/apollo-server/issues/4103 and put a comment in about it within 52418a4.